document contract keys in the ledger model (#2144)
digital-asset/daml: docs/source/concepts/ledger-model/ledger-structure.rst

   The action also contains **actors**, the parties who fetch the contract.
   A **Fetch** behaves like a non-consuming exercise with no consequences, and can be repeated.

> #. a **Key assertion** records the assertion that the given :ref:`contract key <daml-ref-contract-keys>` is not assigned to any active contract on the ledger.

Reviewer: ```suggestion
#. a **Key assertion**, which records the assertion that the given :ref:`contract key <daml-ref-contract-keys>` is not assigned to any active contract on the ledger.
```

Author: Added the `which`.

Author: Replaced `active` with `non-consumed`.

An **Exercise** or a **Fetch** action on a contract is said to **use** the contract.
Moreover, a consuming **Exercise** is said to **consume** (or **archive**) its contract.

The following EBNF-like grammar summarizes the structure of actions
and transactions. Here, "s | t" represents the choice between `s` and

> `t`, "s t" represents `s` followed by `t` and "s*" represents the

Reviewer: I think the serial comma was useful here, as the list items are long and some of them contain "and", so it helps disambiguate.

Author: You're right. I inadvertently deleted the comma.

   The action also contains **actors**, the parties who fetch the contract.
   A **Fetch** behaves like a non-consuming exercise with no consequences, and can be repeated.

> #. a **Key assertion**, which records the assertion that the given :ref:`contract key <daml-ref-contract-keys>` is not assigned to any unconsumed contract on the ledger.

Reviewer: ```suggestion
#. a **Key assertion**, which records the assertion that the given :doc:`contract key </daml/reference/contract-keys>` is not assigned to any unconsumed contract on the ledger.
```

   Action       ::= 'Create' contract
                    | 'Exercise' party* contract Kind Transaction
                    | 'Fetch' party* contract

>                     | 'NoKey' key

Reviewer: Why is this NoKey? It's not clear to me what this means.

Reviewer 2: Perhaps `NoSuchKey`, or `IsFree`? I know it might be painful to rename now.

Author: I like `NoSuchKey`. Renamed. `IsFree` seems too general as it does not refer to a key.

In the paint offer example, the underlying type of contracts consists
of three sorts of contracts:

PaintOffer houseOwner painter obligor refNo

>   Intuitively an offer with a reference number by

Reviewer: In brackets just as it's not the most important detail at this point
```suggestion
  Intuitively an offer (with a reference number) by
```

Author: Done.

   :align: center
   :width: 100%

> To illustrate contract keys, suppose that the contract key for a `PaintOffer` consists of the reference number and the painter as the key maintainer.

Reviewer: It's been a while since this was first mentioned, so I think it's worth having the link to what these are again.
```suggestion
To illustrate :ref:`contract keys <daml-ref-contract-keys>`, suppose that the contract key for a `PaintOffer` consists of the reference number and the painter as the key maintainer.
```

Reviewer 2: The concept of a maintainer hasn't been introduced yet, and I don't think you need it at this point.

Author: Added the link.

Author: Moved the maintainers to the integrity section, but kept the notation with `@` and `&`.

> So Alice can refer to the `PaintOffer` by its key `(P, P123)`.

Reviewer: Is this deliberately using single backticks rather than double?  I guess that is what we do elsewhere in this section, though the rest of the docs would use double backticks (and get code formatting). But probably too much effort to update this entire section.

Author: It's consistent with how it is done in this section, so I'll leave it for now. If you think that it's important to change everything to double backticks, then let's do that as part of another ticket.

> To make this explicit, we use the notation `PaintOffer @P A &P123` for contracts, where `@` marks the maintainers and `&` the other key parts.

Reviewer 2: The `&` notation doesn't seem to be actually used anywhere in the document; I'd drop it.

Author: Have you looked at at the figures? They're full of `&`.

Author: As discussed, this is mainly for visualisation in the diagrams, just like $ for signatories and underlines for observers.

The ledger integrity constraints in the next section ensure that there is always at most one active `PaintOffer` for a given key.

> So if the painter retracts its `PaintOffer` and later Alice tries to accept it, she can then record the absence with a `NoKey (P, P123)` key assertion.

Reviewer: Can she? I would have thought she's not able to, as she's not the maintainer.

Author: The restriction that the submitter is a maintainer is only a limitation of the currently available implementations. It's nothing that we want to enforce on the abstract ledger model. For if we eventually figure out how to do disclosure properly, then the restriction will go away.

   :align: center
   :width: 100%

> To illustrate :ref:`contract keys <daml-ref-contract-keys>`, suppose that the contract key for a `PaintOffer` consists of the reference number and the painter.

Reviewer: ```suggestion
To illustrate :doc:`contract keys </daml/reference/contract-keys>`, suppose that the contract key for a `PaintOffer` consists of the reference number and the painter.
```

   Ledger   ::= Commit*

A DA ledger thus represents the full history of all actions taken by

> parties.\ [#ledger-vs-journal]_ Since the ledger is a sequence (of dependent actions), it induces an

Reviewer: What's this change? The footnote seemed to work fine before, or am I missing something

Author: There was a space between the `.` and the footnote mark. With the `\`, the footnote mark sits directly after the full stop.

   Painter stealing Alice's IOU. Note that the ledger would be
   intuitively permissible if it was Alice performing the last commit.

> .. figure:: ./images/failed-key-assertion.svg

Reviewer 2: I would remove the maintainer annotations from these examples.

Author: So you'd write this as `PaintOffer A &P Bank &P123`? But then we have to change the `&` notation in the privacy section later.

Author: We keep the & and @ as discussed.
